Userland/Games/ColorLines/ColorLines.cpp

```diff
@@ -0,0 +1,402 @@
+    }
+
+    // Draw moving marble
+    if (!m_marble_path.is_empty()) {
+        auto const point = m_marble_path.current_point();
+        auto const& destination_rect = cell_rect.translated(
+            point.x() * board_cell_dimension.width(),
+            point.y() * board_cell_dimension.height());
+        paint_cell(painter, destination_rect, m_board->selected_marble().color(), m_marble_animation_frame);
+    }
+
+    // Draw removing marble
+    if (m_game_state == GameState::MarblesRemoving)
+        for (auto const& marble : m_board->removed_marbles()) {
+            auto const& point = marble.position();
+            auto const& destination_rect = cell_rect.translated(
+                point.x() * board_cell_dimension.width(),
+                point.y() * board_cell_dimension.height());
+            paint_cell(painter, destination_rect, marble.color(), m_marble_animation_frame);
+        }
+
+    // Draw marble move trace
+    if (m_game_state == GameState::MarbleMoving && m_marble_path.remaining_steps() > 1) {
+        auto const trace_size = Gfx::IntSize { m_trace_bitmaps.first()->width(), m_trace_bitmaps.first()->height() };
+        auto const target_trace_size = Gfx::IntSize { 14, 14 };
+        auto const source_rect = Gfx::FloatRect(Gfx::IntPoint {}, trace_size);
+        for (size_t i = 0; i < m_marble_path.remaining_steps() - 1; ++i) {
+            auto const& current_step = m_marble_path[i];
+            auto const destination_rect = Gfx::IntRect(frame_inner_rect().top_left(), target_trace_size)
+                                              .translated(
+                                                  current_step.x() * board_cell_dimension.width(),
+                                                  board_vertical_margin + current_step.y() * board_cell_dimension.height())
+                                              .translated(
+                                                  (board_cell_dimension.width() - target_trace_size.width()) / 2,
+                                                  (board_cell_dimension.height() - target_trace_size.height()) / 2);
+            auto get_direction_bitmap_index = [&]() -> size_t {
+                auto const& previous_step = m_marble_path[i + 1];
+                if (previous_step.x() > current_step.x())
+                    return 3;
+                if (previous_step.x() < current_step.x())
+                    return 1;
+                if (previous_step.y() > current_step.y())
+                    return 0;
+                return 2;
+            };
+            painter.draw_scaled_bitmap(destination_rect, *m_trace_bitmaps[get_direction_bitmap_index()], source_rect,
+                1.0f, Gfx::Painter::ScalingMode::BilinearBlend);
+        }
+    }
+}
+
+void ColorLines::restart_timer(int milliseconds)
+{
+    stop_timer();
+    start_timer(milliseconds);
+}
+
+void ColorLines::set_game_state(GameState state)
+{
+    switch (state) {
```

Comment:
All the branches call
```
m_game_state = state;
update();
```

So it makes sense to do those outside of the `switch`.

Reply:
`m_game_state = state;` can be extracted from all switch cases, i will do that.  The`update()` call is not that uniform, it is called differently in every case, after some case-specific adjustments are performed. So will leave `update()` call in every case as is.